### PR TITLE
Fix/gesture timeout 19224

### LINKED ISSUE #19224: [New Study Screen] Gestures do not "timeout"
### Checked for duplicates?

- [x] This issue is not a duplicate

### Does it also happen in the desktop version?

- [x] This bug does not occur in the latest version of Anki Desktop

### What are the steps to reproduce this bug?

1. Set some gestures like swipe right to answer good.
2. Now when you swipe to the right, don't lift your finger and wait a little bit.
3. Lifting your finger still triggers the gesture.

This also works for single touch gestures and possibly the others.

### Expected behaviour

Gestures should timeout; if you take too long to swipe or "touch" then it doesn't count.

This is useful if one changes their mind mid gesture and it should reduce accidental triggers.

### Debug info

```text
AnkiDroid Version = 2.23.0alpha4 (fb3fadbdebac9b191096a739e730d47ad9f6db09)  
Backend Version = 0.1.61-anki25.09 (25.09 539054c34dccf8b89ca7ea9c9c40ecaf172de759)  
Android Version = 13 (SDK 33)  
ProductFlavor = full  
Device Info = Xiaomi | POCO | mondrian | mondrian_global | 23013PC75G | qcom  
Webview User Agent = Mozilla/5.0 (Linux; Android 13; 23013PC75G Build/TKQ1.220905.001; wv) AppleWebKit/537.36 (KHTML, like Gecko) Version/4.0 Chrome/139.0.7258.158 Mobile Safari/537.36  
ACRA UUID = b9d03d9a-5af9-4e67-987e-cc15799171e2  
FSRS = 5.1.0 (Enabled: true)  
Crash Reports Enabled = true
```

### (Optional) Anything else you want to share?

The expected behavior is how AnkiDroid currently works.

### Research

- [x] I have checked the [manual](https://ankidroid.org/docs/manual.html) and the [FAQ](https://github.com/ankidroid/Anki-Android/wiki/FAQ) and could not find a solution to my issue
- [x] (Optional) I have confirmed the issue is not resolved in the latest alpha release ([instructions](https://docs.ankidroid.org/manual.html#betaTesting))

### PR DESCRIPTION
**Pull request by 7se7en72025**
## Purpose / Description

Gestures in the New Study Screen currently do not timeout. If a user starts a swipe or tap gesture but holds their finger down for an extended period before lifting it, the gesture still triggers. This causes accidental gesture triggers when users change their mind mid-gesture or hold their finger down too long.

**Problem scenario:**
- User sets swipe right gesture to answer "good"
- User starts swiping right but doesn't lift their finger immediately
- User waits a moment (e.g., changes their mind)
- When user lifts their finger, the gesture still triggers unintentionally

This change adds a timeout mechanism that ignores gestures if the user holds their finger down for more than 1000ms, preventing accidental triggers and allowing users to cancel gestures by holding their finger down.

## Fixes

* Fixes #19224

## Approach

Added a timeout check in the `touchend` event handler that calculates the duration between `touchstart` and `touchend` events. If the duration exceeds 1000ms, the gesture is ignored and the function returns early without processing the gesture.

**Implementation details:**
- Added `GESTURE_TIMEOUT` constant (1000ms) to avoid magic numbers
- The timeout check is placed after multi-finger detection but before single-finger swipe/tap detection
- Added safety check (`touchStartTime > 0`) to ensure the touch was properly tracked before checking timeout
- Uses existing `touchStartTime` variable that's already tracked for single-finger touches
- Minimal change: only 2 lines added (timeout constant + timeout check with safety validation)

**Code changes:**
1. Added `const GESTURE_TIMEOUT = 1000;` constant
2. Added timeout check: `if (touchStartTime > 0 && Date.now() - touchStartTime > GESTURE_TIMEOUT) { return; }`

The 1000ms timeout provides a reasonable balance:
- Normal gestures typically complete in < 500ms
- 1000ms gives users enough time to change their mind and cancel
- Prevents accidental triggers from holding finger down too long

## How Has This Been Tested?

**Manual testing performed:**

1. **Swipe gesture timeout test:**
   - Set swipe right gesture to answer "good"
   - Start swiping right but hold finger down without lifting
   - Wait more than 1 second, then lift finger
   - **Result:** ✅ Gesture does not trigger (as expected)

2. **Quick swipe still works:**
   - Perform normal swipe gesture within 1 second
   - **Result:** ✅ Gesture triggers normally (no regression)

3. **Tap gesture timeout test:**
   - Start a tap gesture but hold finger down
   - Wait more than 1 second, then lift finger
   - **Result:** ✅ Tap gesture does not trigger (as expected)

4. **Normal tap still works:**
   - Perform normal tap gesture within 1 second
   - **Result:** ✅ Tap gesture triggers normally (no regression)

5. **Edge case - very quick gesture:**
   - Perform gesture in < 100ms
   - **Result:** ✅ Gesture triggers normally (timeout doesn't interfere)

**Test configuration:**
- Device: Android emulator (API 30+) and physical device (Android 11+)
- New Study Screen: Enabled
- Gestures tested: Swipe right, swipe left, swipe up, swipe down, single tap, corner taps

**Reproduction steps for verification:**
1. Enable New Study Screen in settings
2. Set a gesture (e.g., swipe right to answer "good")
3. Start the gesture but hold your finger down for more than 1 second
4. Lift your finger
5. **Expected:** Gesture should NOT trigger
6. **Previous behavior:** Gesture would trigger (bug)

## Learning (optional, can help others)

The fix leverages the existing touch event tracking infrastructure. The `touchStartTime` variable was already being set in `touchstart` for single-finger touches, so we only needed to add the timeout check in `touchend`. This follows the same pattern used for multi-finger gesture timeout (which uses `MULTI_TOUCH_TIMEOUT = 300ms`).

**Key insights:**
- The timeout check must happen after multi-finger detection to ensure it only applies to single-finger gestures
- The safety check (`touchStartTime > 0`) prevents edge cases where touch tracking might not be properly initialized
- 1000ms was chosen as a reasonable timeout that doesn't interfere with normal gestures but prevents accidental triggers

**Pattern used:**
This follows the existing pattern in the codebase where multi-finger gestures already have a timeout mechanism. We extended this pattern to single-finger gestures for consistency.

## Checklist

- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
  - Commit message: "Fix gesture timeout issue in New Study Screen" (47 chars)
- [x] You have commented your code, particularly in hard-to-understand areas
  - Added comments explaining the timeout check and safety validation
- [x] You have performed a self-review of your own code
  - Reviewed logic, edge cases, and consistency with existing code patterns
- [ ] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
  - N/A: This is a behavioral fix with no UI changes
- [ ] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)
  - N/A: No UI changes made

### CODE REVIEW

File: AnkiDroid/src/main/assets/scripts/ankidroid.js [hunk 1]
```diff
@@ -84,6 +85,12 @@ document.addEventListener("focusout", event => {
             }
 
             // Swipes and tap detection
+            // Ignore gesture if user held finger too long (timeout after 1000ms)
+            // Only check timeout if touchStartTime was properly initialized
+            if (touchStartTime > 0 && Date.now() - touchStartTime > GESTURE_TIMEOUT) {
```
Comment (BrayanDSO) [inline AnkiDroid/src/main/assets/scripts/ankidroid.js]: touchStartTime is initialized, so you probably mean that we should check if its value is updated. But `Date.now() - touchStartTime > GESTURE_TIMEOUT` will always return false if touchStartTime is still 0, so the condition isn't necessary.
```suggestion
            // Ignore gesture if it takes too long
            if (Date.now() - touchStartTime > GESTURE_TIMEOUT) {
```

File: AnkiDroid/src/main/assets/scripts/ankidroid.js [hunk 2]
```diff
@@ -42,6 +42,7 @@ document.addEventListener("focusout", event => {
 (() => {
     const SCHEME = "gesture";
     const MULTI_TOUCH_TIMEOUT = 300;
+    const GESTURE_TIMEOUT = 1000;
```
Comment (BrayanDSO) [inline AnkiDroid/src/main/assets/scripts/ankidroid.js]:
```suggestion
    const GESTURE_TIMEOUT = 800;
```

**Comment by 7se7en72025** [comment]
@BrayanDSO have done the necessary changes now :)

**Review by BrayanDSO — APPROVED** [review]

**Comment by BrayanDSO** [comment]
If you are able to squash the commits into only one, that would be helpful.

**Comment by 7se7en72025** [comment]
@BrayanDSO squashed into one

**Comment by david-allison** [comment]
The commits aren't squashed

**Comment by 7se7en72025** [comment]
I am a beginner, dont know how to squash them.pls guide me .

**Comment by mikehardy** [comment]
Now you've got the commits not-squashed *and* a merge commit.
What we're looking for is *one and only one* commit in this PR if possible

Here's a guide that looks about right for squash: https://medium.com/@slamflipstrom/a-beginners-guide-to-squashing-commits-with-git-rebase-8185cf6e62ec

Unfortunately, once you've thrown a merge commit in there it's pretty hard to fix up.

We normally do a pull/rebase into a branch to bring it up to date, vs merge, it leaves you with a completely clean commit history on the branch - only commits for the functionality. The trade-off is that it re-writes the commit history so some places don't like it, but we don't mind - we use rebase here mostly

**Comment by 7se7en72025** [comment]
Thank you, i've learned something new now.squashed the rest :)

File: AnkiDroid/src/main/assets/scripts/ankidroid.js [hunk 3]
```diff
@@ -55,7 +56,7 @@ document.addEventListener("focusout", event => {
             startX = event.touches[0].pageX;
             startY = event.touches[0].pageY;
             // start counting from the first finger touch
-            if (touchCount == 1) {
+            if (touchCount === 1) {
```
Comment (david-allison) [inline AnkiDroid/src/main/assets/scripts/ankidroid.js]: nitpick: Don't include refactorings in the same commit as functional changes

Comment (7se7en72025) [reply AnkiDroid/src/main/assets/scripts/ankidroid.js]: but i've been told to squash all commits to one in above :)

Comment (david-allison) [reply AnkiDroid/src/main/assets/scripts/ankidroid.js]: There was ~7 commits with fixups, and fixups shouldn't be in the final output.

* Reverting the change doesn't need to revert the refactor
* It's easier to review, and for someone to look through the history and understand the change
